rust/onnxruntime/src/session.rs

```diff
@@ -0,0 +1,741 @@
+    session_ptr: *mut sys::OrtSession,
+    allocator_ptr: *mut sys::OrtAllocator,
+    memory_info: MemoryInfo,
+    /// Information about the ONNX's inputs as stored in loaded file
+    pub inputs: Vec<Input>,
+    /// Information about the ONNX's outputs as stored in loaded file
+    pub outputs: Vec<Output>,
+}
+
+/// Information about an ONNX's input as stored in loaded file
+#[derive(Debug)]
+pub struct Input {
+    /// Name of the input layer
+    pub name: String,
+    /// Type of the input layer's elements
+    pub input_type: TensorElementDataType,
+    /// Shape of the input layer
+    ///
+    /// C API uses a i64 for the dimensions. We use an unsigned of the same range of the positive values.
+    pub dimensions: Vec<Option<u32>>,
+}
+
+/// Information about an ONNX's output as stored in loaded file
+#[derive(Debug)]
+pub struct Output {
+    /// Name of the output layer
+    pub name: String,
+    /// Type of the output layer's elements
+    pub output_type: TensorElementDataType,
+    /// Shape of the output layer
+    ///
+    /// C API uses a i64 for the dimensions. We use an unsigned of the same range of the positive values.
+    pub dimensions: Vec<Option<u32>>,
+}
+
+impl Input {
+    /// Return an iterator over the shape elements of the input layer
+    ///
+    /// Note: The member [`Input::dimensions`](struct.Input.html#structfield.dimensions)
+    /// stores `u32` (since ONNX uses `i64` but which cannot be negative) so the
+    /// iterator converts to `usize`.
+    pub fn dimensions(&self) -> impl Iterator<Item = Option<usize>> + '_ {
+        self.dimensions.iter().map(|d| d.map(|d2| d2 as usize))
+    }
+}
+
+impl Output {
+    /// Return an iterator over the shape elements of the output layer
+    ///
+    /// Note: The member [`Output::dimensions`](struct.Output.html#structfield.dimensions)
+    /// stores `u32` (since ONNX uses `i64` but which cannot be negative) so the
+    /// iterator converts to `usize`.
+    pub fn dimensions(&self) -> impl Iterator<Item = Option<usize>> + '_ {
+        self.dimensions.iter().map(|d| d.map(|d2| d2 as usize))
+    }
+}
+
+impl Drop for Session {
+    #[tracing::instrument]
+    fn drop(&mut self) {
+        debug!("Dropping the session.");
+        if self.session_ptr.is_null() {
+            error!("Session pointer is null, not dropping.");
+        } else {
+            unsafe { g_ort().ReleaseSession.unwrap()(self.session_ptr) };
+        }
+
+        self.session_ptr = std::ptr::null_mut();
+        self.allocator_ptr = std::ptr::null_mut();
+    }
+}
+
+unsafe impl Send for Session {}
+
+impl Session {
+    /// Run the input data through the ONNX graph, performing inference.
+    ///
+    /// Note that ONNX models can have multiple inputs; a `Vec<_>` is thus
+    /// used for the input data here.
+    pub fn run<T>(&self, input_arrays: Vec<impl ConstructTensor + Debug>) -> Result<Vec<T>>
+    where
+        T: TryFrom<OrtOutputTensor, Error = OrtError> + Debug,
```

Comment:
This method can only handle ONNX models whose output tensors have the same element type, can't it? The outputs of a model may be of different types because ONNX allows the outputs of a model to come from multiple operators, rather than a single (fanning-in) operator. Also, there are several multi-output ONNX operators whose outputs can be of different types: BatchNormalization, Dropout, Loop, MaxPool, etc. That was one of the reasons that FuriosaAI had to create its own Rust bindings, [OnnxRt](https://github.com/furiosa-ai/onnxrt), to ONNX Runtime.

Reply:
Yeah, Thanks. I think I am going to go with an enum for the output, so there can be multiple output types and the user gets to match on which type they have.

Reply:
@dkim-furiosa I changed the last commit to allow for inputs of different types, and outputs of different types.

Reply:
First of all, thank you for your consideration!

Can I ask a question about the new signature of `Session::run`?

```rust
pub fn run<'a>(&self, input_arrays: Vec<&mut dyn ConstructTensor>) -> Result<Vec<OrtOutput<'a>>>
```

The lifetime parameter `'a` appears not to be constrained in any way. If so, is it equivalent to `'static`?

Reply:
Yes, I think so. I think it is probably safer to have it tied to the lifetime of `&self`.

Reply:
@dkim-furiosa Does the Session::run method meet your needs, as included in this PR?

Reply:
Can you consider returning output tensors via an output parameter of `Session::run`? This will enable memory for output tensors to be reused across multiple invocations of `Session::run` of the same model. Currently, each invocation of `Session::run` will perform (heap) memory allocations for output tensors. An output parameter will also allow the lifetime of output tensors to be independent of that of a session, consistent of [the original C interface](https://github.com/microsoft/onnxruntime/blob/e01a8519e0a2a3550f0e0edd609bb5058912b6a9/include/onnxruntime/core/session/onnxruntime_c_api.h#L710-L732). You might be interested to refer to the interface of OnnxRt's [`Session::run`](https://github.com/furiosa-ai/onnxrt/blob/9a37030890d67039bff2506f0ac2fd742e68af57/src/lib.rs#L1247-L1254).

Reply:
Thanks. I will look into it.

Reply:
Hi @dkim-furiosa, I don't want to use out parameters because it will significantly change the function signature. I want the signature to be similar to what it has been so that people who have code based on the old bindings won't have to update too much.

Reply:
Is there any way to add a second function that allows this? We have many users that supply the output tensors for re-use as it's much more efficient especially when doing streaming style processing (video frames/audio/etc).

Reply:
Thanks for the context, @RyanUnderhill . I will look into a second function that allows run with out parameters.